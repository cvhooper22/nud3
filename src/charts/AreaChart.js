import React, { Component, PropTypes } from 'react';
import * as d3 from 'd3';
import _ from 'lodash';

const stringOrArrayOfStrings = PropTypes.oneOfType([
  PropTypes.string,
  React.PropTypes.arrayOf(PropTypes.string),
]);

export default class AreaChart extends Component {

  static propTypes = {
    chartData: PropTypes.array,
    className: PropTypes.string,
    classNamePrefix: PropTypes.string,
    colorPalette: PropTypes.any,
    filter: stringOrArrayOfStrings,
    paddingBottom: PropTypes.number,
    paddingLeft: PropTypes.number,
    paddingRight: PropTypes.number,
    paddingTop: PropTypes.number,
    valueKeys: PropTypes.array,
    xScale: PropTypes.func,
    yScale: PropTypes.func,
    height: PropTypes.number,
    areaHeight: PropTypes.number,
    clipPath: PropTypes.string,
  };

  constructor (...args) {
    super(...args);
    this.getUniqueDataKey = ::this.getUniqueDataKey;
    this.getFillColor = ::this.getFillColor;
    this.area = d3.area()
      .x(d => this.props.xScale(d.xValue))
      .y1(d => this.props.yScale(d.yValue || 0))
      .y0(() => this.props.areaHeight);
  }

  componentDidMount () {
    this.renderArea();
  }

  componentDidUpdate () {
    this.renderArea();
  }

  getFillColor (d, i) {
    if (_.isFunction(this.props.colorPalette)) {
      return this.props.colorPalette(i);
    }
    return this.props.colorPalette[i];
  }

  getUniqueDataKey (dataSet, i) {
    return this.props.valueKeys[i];
  }

  getPathFilter (d, i) {
    let filter = this.props.filter;
    if (_.isArray(filter)) {
      filter = filter[i];
    }
    return `url(#${filter})`;
  }

  render () {
    const className = [
      'area-chart',
    ];
    if (this.props.classNamePrefix) {
      className.push(`${this.props.classNamePrefix}__area-chart`);
    }
    if (this.props.className) {
      className.push(this.props.className);
    }
    return (
      <g
        className={ className.join(' ') }
        transform={ `translate(${this.props.paddingLeft},${this.props.paddingTop})` }
        clipPath={ this.props.clipPath }
        ref={ n => this.node = n }
      />
    );
  }

  renderArea () {
    this.group = d3.select(this.node);
    const paths = this.group.selectAll('.area-chart__area')
      .data(this.props.chartData, this.getUniqueDataKey)
      .enter()
        .append('path');
    paths
      .attr('class', 'area-chart__area');
    if (this.props.filter) {
      paths.style('filter', ::this.getPathFilter);
    }

    const areas = this.group
      .selectAll('.area-chart__area')
      .attr('d', this.area)
    if (this.props.colorPalette) {
      areas.style('fill', this.getFillColor);
    }
  }
}
