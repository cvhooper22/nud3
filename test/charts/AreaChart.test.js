import React from 'react';
import renderer from 'react-test-renderer';
import AreaChart from '../../src/charts/AreaChart';
import {
  chartData,
  yScale,
  xScale,
  valueKeys,
  originalData,
} from '../mockData/chartData';

describe('AreaChart.js', () => {
  it('should take stubbed data and do basic rendering', () => {
    const tree = renderer.create(
      <AreaChart
        originalData={ originalData }
        chartData={ chartData }
        areaHeight={ 100 }
        areaWidth={ 100 }
        paddingTop={ 0 }
        paddingBottom={ 0 }
        paddingRight={ 0 }
        paddingLeft={ 0 }
        valueKeys={ valueKeys }
        yScale={ yScale }
        xScale={ xScale }
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('should take stubbed data repositions based on padding', () => {
    const tree = renderer.create(
      <AreaChart
        originalData={ originalData }
        chartData={ chartData }
        areaHeight={ 100 }
        areaWidth={ 100 }
        paddingTop={ 10 }
        paddingBottom={ 10 }
        paddingRight={ 10 }
        paddingLeft={ 10 }
        valueKeys={ valueKeys }
        yScale={ yScale }
        xScale={ xScale }
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('should take the clip path prop', () => {
    const tree = renderer.create(
      <AreaChart
        originalData={ originalData }
        chartData={ chartData }
        areaHeight={ 100 }
        areaWidth={ 100 }
        paddingTop={ 10 }
        paddingBottom={ 10 }
        paddingRight={ 10 }
        paddingLeft={ 10 }
        valueKeys={ valueKeys }
        yScale={ yScale }
        xScale={ xScale }
        clipPath="my-clip-path"
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
