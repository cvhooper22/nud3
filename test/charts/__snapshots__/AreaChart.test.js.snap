exports[`AreaChart.js should take stubbed data and do basic rendering 1`] = `
<g
  className="area-chart"
  clipPath={undefined}
  transform="translate(0,0)" />
`;

exports[`AreaChart.js should take stubbed data repositions based on padding 1`] = `
<g
  className="area-chart"
  clipPath={undefined}
  transform="translate(10,10)" />
`;

exports[`AreaChart.js should take the clip path prop 1`] = `
<g
  className="area-chart"
  clipPath="my-clip-path"
  transform="translate(10,10)" />
`;
